harness/determined/common/api/authentication.py

```diff
@@ -38,204 +26,171 @@ def __init__(self, username: str, token: str):
-        # user and we ended up falling back to the default username `determined`, then we can fall
-        # back to the default login as well. Otherwise, ask the user for their password.
-        fallback_to_default = password is None and session_user == constants.DEFAULT_DETERMINED_USER
-        if fallback_to_default:
-            password = constants.DEFAULT_DETERMINED_PASSWORD
-
-        if password is None:
-            password = getpass.getpass("Password for user '{}': ".format(session_user))
-
-        try:
-            token = do_login(self.master_address, session_user, password, cert)
-        except api.errors.ForbiddenException:
-            if fallback_to_default:
-                raise api.errors.UnauthenticatedException(username=session_user)
-            raise
-
-        self.token_store.set_token(session_user, token)
-
-        return UsernameTokenPair(session_user, token)
-
-    def is_user_active(self, username: str) -> bool:
-        return self.token_store.get_active_user() == username
+    env_user = util.get_det_username_from_env()
+    env_pass = util.get_det_password_from_env()
+    if env_user is not None and env_pass is not None:
+        return env_user, env_pass, False
+
+    # Next priority is the active user in the token store.
+    active_user = token_store.get_active_user()
+    if active_user is not None:
+        return active_user, password, False
+
+    # Last priority is the default username and password.
+    return (
+        constants.DEFAULT_DETERMINED_USER,
+        password or constants.DEFAULT_DETERMINED_PASSWORD,
+        True,
+    )
+
+
+def login_with_cache(
+    master_address: Optional[str] = None,
+    requested_user: Optional[str] = None,
+    password: Optional[str] = None,
+    cert: Optional[certs.Cert] = None,
+) -> UsernameTokenPair:
+    """
+    Log in, preferring cached credentials in the TokenStore, if possible.
 
-    def get_session_user(self) -> str:
-        """
-        Returns the session user for the current session. If there is no active
-        session, then an UnauthenticatedException will be raised.
-        """
-        return self.session.username
+    This is the login path for nearly all user-facing cases.
 
-    def get_session_token(self, must: bool = True) -> str:
-        """
-        Returns the authentication token for the session user. If there is no
-        active session, then an UnauthenticatedException will be raised.
-        """
-        if self.session is None:
-            if must:
-                raise api.errors.UnauthenticatedException(username="")
-            else:
-                return ""
-        return self.session.token
+    There is also a special case for checking if the DET_USER_TOKEN is set in the environment (by
+    the determined-master).  That must happen in this function because it is only used when no other
+    login tokens are active, but it must be considered before asking the user for a password.
```

Comment:
This is a place where I think a `Returns` section would be helpful. In general, I'd like the docs to tell me what this function actually does.

I may not be the smartest developer, but I've really struggled with what "log in" means both before this PR and during this code review.

Reply:
done